config/204-psp.yaml

```diff
@@ -0,0 +1,36 @@
+# Copyright 2019 The Knative Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: extensions/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  name: knative-serving-psp
+  labels:
+    serving.knative.dev/release: devel
+    serving.knative.dev/controller: "true"
+spec:
+  allowPrivilegeEscalation: true
+  allowedCapabilities:
+  - 'NET_ADMIN'
+  fsGroup:
+    rule: RunAsAny
+  privileged: true
+  runAsUser:
+    rule: RunAsAny
+  seLinux:
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
```

Comment:
Do these things need to be as permissive as they are given that we've stopped running as root?

Reply:
Because these fields are mandatory, without them, PSP couldn't created.
Without them following errors will be reported
```
error: error validating "psp.yaml": error validating data: [ValidationError(PodSecurityPolicy.spec): missing required field "seLinux" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec, ValidationError(PodSecurityPolicy.spec): missing required field "runAsUser" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec, ValidationError(PodSecurityPolicy.spec): missing required field "supplementalGroups" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec, ValidationError(PodSecurityPolicy.spec): missing required field "fsGroup" in io.k8s.api.extensions.v1beta1.PodSecurityPolicySpec]; if you choose to ignore these errors, turn validation off with --validate=false
```

Reply:
I'm pretty new to PSP, but this looks like it allows us to do anything, so I'm curious how this is an improvement over where we were prior to `:nonroot`?

Reply:
Actually, if PSP is used, :nonroot images is not necessary any more.
Another choice is hand over PSP related stuff to cluster administrator to handle.

Reply:
Now I am _more_ confused :)

If PSP validates the `USER` then shouldn't it block our controllers, which run as root (without `:nonroot`)?

Maybe the answer is that I need to read what this does...

Reply:
If PSP validates user and container needs to run as root, it will be blocked.
For example, I created psp for Knative serving as below,
```
root@filmy1:/mnt/data/gowork/src/github.com/knative/serving# knative get psp knative-serving-psp
NAME                  PRIV   CAPS        SELINUX    RUNASUSER          FSGROUP    SUPGROUP   READONLYROOTFS   VOLUMES
knative-serving-psp   true   NET_ADMIN   RunAsAny   MustRunAsNonRoot   RunAsAny   RunAsAny   false            *
```
Then I build knative serving with base images ubuntu which need to run as root.
Pod failed to start as below,
```
  Normal   Pulling    3m34s                 kubelet, 172.16.17.35  pulling image "icpdev.icp:8500/imgrepo/controller-43f0364ab2f6dab17267e80a1f6a4adc@sha256:4806fc2cc89330d8731385033012f4f04ebc43d5f4918af999b7528162e9dd2b"
  Normal   Pulled     3m30s                 kubelet, 172.16.17.35  Successfully pulled image "icpdev.icp:8500/imgrepo/controller-43f0364ab2f6dab17267e80a1f6a4adc@sha256:4806fc2cc89330d8731385033012f4f04ebc43d5f4918af999b7528162e9dd2b"
  Warning  Failed     76s (x12 over 3m30s)  kubelet, 172.16.17.35  Error: container has runAsNonRoot and image will run as root
  Normal   Pulled     76s (x11 over 3m29s)  kubelet, 172.16.17.35  Container image "icpdev.icp:8500/imgrepo/controller-43f0364ab2f6dab17267e80a1f6a4adc@sha256:4806fc2cc89330d8731385033012f4f04ebc43d5f4918af999b7528162e9dd2b" already present on machine
```

Reply:
If Istio auto injection is enabled, it needs capability of 'NET_ADMIN' and needs to be run as root. And RunAsUser in psp should be set to RunAsAny.

Reply:
Ok, so `tl;dr` this is the PSP one needs with Istio Mesh, which is scary.

What I'm wondering is what **knative/serving** needs when run _without_ sidecars, e.g. istio-lean.yaml?

My hope was for us to be less scary (and use PSP to verify that), but I see little value in testing it in scenarios where the mesh potentially masks all of our work.  If we end up needing two PSPs (mesh and no-mesh) that's fine, and we can tweak scripts to install the right ones.

thanks for your patience while I wrap my head around this.